pkg/requests/builder.go

```diff
@@ -0,0 +1,173 @@
+package requests
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/bitly/go-simplejson"
+)
+
+// Builder allows users to construct a request and then either get the requests
+// response via Do(), parse the response into a simplejson.Json via JSON(),
+// or to parse the json response into an object via Into().
+type Builder interface {
+	WithContext(context.Context) Builder
+	WithBody(io.Reader) Builder
+	WithMethod(string) Builder
+	WithHeader(http.Header) Builder
+	SetHeader(key, value string) Builder
+	Do() (*http.Response, error)
+	Into(interface{}) error
+	JSON() (*simplejson.Json, error)
+}
+
+type builder struct {
+	context  context.Context
+	method   string
+	endpoint string
+	body     io.Reader
+	header   http.Header
+	response *http.Response
+}
+
+// New provides a new Builder for the given endpoint.
+func New(endpoint string) Builder {
+	return &builder{
+		endpoint: endpoint,
+		method:   "GET",
+	}
+}
+
+// WithContext adds a context to the request.
+// If no context is provided, context.Background() is used instead.
+func (r *builder) WithContext(ctx context.Context) Builder {
+	r.context = ctx
+	return r
+}
+
+// WithBody adds a body to the request.
+func (r *builder) WithBody(body io.Reader) Builder {
+	r.body = body
+	return r
+}
+
+// WithMethod sets the request method. Defaults to "GET".
+func (r *builder) WithMethod(method string) Builder {
+	r.method = method
+	return r
+}
+
+// WithHeader replaces the request header map with the given header map.
+func (r *builder) WithHeader(header http.Header) Builder {
+	r.header = header
+	return r
+}
+
+// SetHeader sets a single header to the given value.
+// May be used to add multiple headers.
+func (r *builder) SetHeader(key, value string) Builder {
+	if r.header == nil {
+		r.header = make(http.Header)
+	}
+	r.header.Set(key, value)
+	return r
+}
+
+// Do performs the request and returns the response in its raw form.
+// If the request has already been performed, returns the previous result.
+// This will not allow you to repeat a request.
+func (r *builder) Do() (*http.Response, error) {
```

Comment:
I'm torn on this: now after looking I see the naming aligns to `http.DefaultClient.Do(req)` -- but on a cursory glance at the PR the purpose of this function didn't jump out to me by eyeballing the name.

"Do... what?" was my first thought.  Is `Send` potentially more clear?

Reply:
I think this becomes move obvious when it's used, `requests.New(url.String()).Do()` is a pattern I've seen before, it was somewhere loosely based on the Kubernetes REST client  https://github.com/kubernetes/client-go/blob/1e6150831a960cf38075bb8561e01f48837f8d74/rest/request.go#L952

`Send` to me feels a bit wrong 😅 It sends but also receives the result of the request 🤔

Reply:
It looks like `Do` seems to be an overwhelming convention in this space -- lets stick with that! Today I Learned 😄